portal-web/test/functional/com/liferay/portalweb/macros/SAML.macro

```diff
@@ -1,5 +1,125 @@
+				User.loginPG(
+					password = "test",
+					specificURL = "${spURL}",
+					userEmailAddress = "${spAdminEmailAddress}");
+			}
+			else {
+				User.loginPG(
+					password = "test",
+					specificURL = "${spURL}",
+					userEmailAddress = "test@liferay.com");
+			}
+		}
+
+		ApplicationsMenu.gotoPortlet(
+			category = "Security",
+			panel = "Control Panel",
+			portlet = "SAML Admin");
+
+		SAMLNavigation.gotoIdentityProviderConnections();
+
+		SAMLPortlet.addIdentityProviderConnection(
+			idpAttributes = "${idpAttributes}",
+			idpEntityID = "${idpEntityID}",
+			idpKeepAliveURL = "${idpKeepAliveURL}",
+			idpMetadataURL = "${idpMetadataURL}/c/portal/saml/metadata",
+			idpNameID = "${idpEntityID}",
+			nameIdentifierFormat = "${nameIdentifierFormat}");
+	}
+
+	@description = "Add new SP Connection data for IDP."
+	macro addNewSPConnection {
+		Navigator.openSpecificURL(url = "${idpURL}");
+
+		if (IsElementNotPresent(locator1 = "UserBar#USER_AVATAR_IMAGE")) {
+			User.loginPG(
+				password = "test",
+				specificURL = "${idpURL}",
+				userEmailAddress = "test@liferay.com");
+		}
+
+		ApplicationsMenu.gotoPortlet(
+			category = "Security",
+			panel = "Control Panel",
+			portlet = "SAML Admin");
+
+		SAMLNavigation.gotoServiceProviderConnection();
+
+		spEntityID = "${spEntityID}";
+		spMetadataURL = "${spURL}/c/portal/saml/metadata";
+		spNameID = "${spEntityID}";
+
+		AssertClick(
+			locator1 = "CPSAMLAdminServiceProviderConnections#ADD_SERVICE_PROVIDER_BUTTON",
+			value1 = "Add Service Provider");
+
+		Type(
+			locator1 = "TextInput#SAML_NAME_FIELD",
+			value1 = "${spNameID}");
+
+		Type(
+			locator1 = "CPSAMLAdminServiceProviderConnections#SERVICE_PROVIDER_CONNECTION_ENTITY_ID_FIELD",
+			value1 = "${spEntityID}");
+
+		Check(locator1 = "CPSAMLAdminServiceProviderConnections#SERVICE_PROVIDER_CONNECTION_ENABLED_CHECKBOX");
+
+		Type(
+			locator1 = "TextInput#SAML_METADATA_URL_FIELD",
+			value1 = "${spMetadataURL}");
+
+		if (!(isSet(nameIdentifierAttributeName))) {
+			var nameIdentifierAttributeName = "emailAddress";
+		}
+
+		Type(
+			locator1 = "CPSAMLAdminServiceProviderConnections#SERVICE_PROVIDER_CONNECTION_NAME_IDENTIFIER_ATTRIBUTE_NAME_FIELD",
+			value1 = "${nameIdentifierAttributeName}");
+
+		if (isSet(nameIdentifierFormat)) {
+			Select(
+				locator1 = "Select#SAML_NAME_IDENTIFIER_FORMAT_SELECT",
+				value1 = "${nameIdentifierFormat}");
+		}
+
+		if (isSet(spAttributes)) {
+			Check(locator1 = "CPSAMLAdminServiceProviderConnections#SERVICE_PROVIDER_CONNECTION_ATTRIBUTE_ENABLE");
+
+			Type(
+				locator1 = "CPSAMLAdminServiceProviderConnections#SERVICE_PROVIDER_CONNECTION_ATTRIBUTE_TEXTAREA",
+				value1 = "${spAttributes}");
+		}
+
+		if (isSet(spKeepAliveURL)) {
+			Type(
+				locator1 = "CPSAMLAdmin#SAML_KEEP_ALIVE_URL",
+				value1 = "${spKeepAliveURL}");
+		}
+
+		PortletEntry.save();
+	}
+
+	@description = "Verify how many times the user exists"
+	macro AssertUserCount {
```

Comment:
Should be `assertUserCount`. Also would we just be able to verify that a second instance of the user doesn't exist ([example](https://stackoverflow.com/questions/4007413/xpath-query-to-get-nth-instance-of-an-element) instead of creating a macro to verify this by search? Or will this macro see more use?